file_transfer_agent.go

```diff
@@ -464,45 +640,54 @@ func (sfa *snowflakeFileTransferAgent) uploadFilesParallel(fileMetas []*fileMeta
 	for idx < fileMetaLen {
 		endOfIdx := intMin(fileMetaLen, idx+int(sfa.parallel))
 		targetMeta := fileMetas[idx:endOfIdx]
-		errChan := make(chan error, len(targetMeta))
 		for {
 			var wg sync.WaitGroup
-			results := make([]*fileMetadata, fileMetaLen)
+			results := make([]*fileMetadata, len(targetMeta))
+			errors := make([]error, len(targetMeta))
 			for i, meta := range targetMeta {
 				wg.Add(1)
-				go func(k int, m *fileMetadata, errChan chan error) {
+				go func(k int, m *fileMetadata) {
 					defer wg.Done()
-					var uploadErr error
-					results[k], uploadErr = sfa.uploadOneFile(m)
-					errChan <- uploadErr
-				}(i, meta, errChan)
+					results[k], errors[k] = sfa.uploadOneFile(m)
```

Comment:
For this async implementation, since we are not using any thread here, if there are 1000 files to upload, in the middle the upload failed because token expires, then what will happen? Seems the client will keep sending all these 1000X12M files, then rely on the logic at the end (line 656) to collect error statuses to retry. Should we try to be more efficient here, e.g. at the beginning of uploadOneFile, check if other execution of this uploadOneFile failed with renewToken or renewPresignedURL, then the current upload should not continue. I would suggest to add a test to mock a time out case, and for GCS, a renewPresignedURL so you could test the code path here more thoroughly and compare the performance.

Reply:
The thought was that we try to upload regardless and if it fails because of an expired client or anything of similar nature where it requires the driver to take action before retrying, we'd try to fix the problem before trying to re-upload again. If we short circuit the upload, wouldn't that lead to not uploading files we otherwise could have?

Reply:
error reported for expired token is something that will fail every upload after the expiration. So this is not random thing. We should pause the uploads that are not started, refresh all clients (these clients are created at the same time so they should be expired at this point) then resume the uploads. This is the logic in other drivers. If you prefer to defer the efficiently handling expired token late I think that is also fine.